packages/matter.js/InterfaceNotes.ts

```diff
@@ -0,0 +1,121 @@
+    // of endpoints and thousands of attributes on a low-power device?
+    //
+    // Note 4: (no longer relevant)
+    set(changes: Partial<State>): void;
+
+    // This is a convenience form of set for updating a single attribute.
+    //
+    // Note 8: We could also generate a setter for every individual attribute.
+    // However, I worry that would encourage inefficient implementation.
+    //
+    // Note 9: Not sure how to validate value type in the interface so might
+    // need to do that programmatically.
+    set(name: keyof (State), value: any): void;
+
+    // This is how matter.js conveys a request to change values from the
+    // fabric.  It would be wired to AttributeServers via InteractionServer
+    // internally.
+    //
+    // Question 4: We discussed whether this should be asynchronous or not.
+    // I'm thinking of use cases like bridges or even I/O in the case of a
+    // local device implementation.  But should this be asynchronous so an
+    // attribute write fails if the "source of truth" rejects the change?
+    //
+    // Question 5: I worry about chattiness here too.  If we don't shape the
+    // API like this, are we potentially forcing API consumers to handle
+    // coalescing of values on their side?  For example, if a controller says
+    // "turn this light to 70%, red, with a 5 second delay" am I correct in
+    // surmising that a Hue bridge implementation would want to take receipt of
+    // all three attribute changes simultaneously?
+    write(changes: Partial<State>): Promise<void>;
+
+    // COMMANDS
+    //
+    // These are callbacks.  The internal device implementation wires these to
+    // a CommandServer via CommandHandler.  This means a device implementer
+    // just needs to override a function to implement the functionality.
+    //
+    // Note 5: We can offer default implementations for these functions that
+    // try to make the user's life easier.  But until we do that we can just
+    // install a default implementation that throws a "not supported" error.
+    // Then the fancier implementation won't be a breaking API change, it'll
+    // just be a smarter default.
+    invokeMoveToLevel(request: MoveToLevelRequest): Promise<void>;
+    invokeMove(request: MoveRequest): Promise<void>;
+    invokeStep(request: StepRequest): Promise<void>;
+    invokeStop(request: StopRequest): Promise<void>;
+    invokeMoveToLevelWithOnOff(request: MoveToLevelWithOnOffRequest): Promise<void>;
+    invokeMoveWithOnOff(request: MoveWithOnOffRequest): Promise<void>;
+    invokeStepWithOnOff(request: StepWithOnOffRequest): Promise<void>;
+    invokeStopWithOnOff(request: StopWithOnOffRequest): Promise<void>;
+
+    // EVENTS
+    //
+    // These are invoked by the API consumer to notify matter.js of an event.
+    //
+    // Question 6: I haven't figured where to wire these yet.  Is this bit of
+    // interaction server still in flight or did I just miss it?
```

Comment:
Ok, I had a look into events ... will be interesting. 

My current thinking is that on "new ClusterServer" we enhance it by a "supportedCommands" object where you define which events you support aka "your code can generate". then clusterServer instance (?) will have a "triggerEventXY" method or we have an EventManager or such internally.

On Client side events will be part or read, so there will be "getXYEvent" methods I assume on clusterclient and an option to add it to a subscribe.
Events alwys might return a list of events happened since last "get" and events of same name could come in that list ...

so for API: TriggerEvent methods to generate an event and for clients "eventListeners"

Reply:
I will add "put in top region" for after vacation because will have bigger impact on API so we should do soon